org.eclipse.ice.dev/org.eclipse.ice.dev.annotations/src/main/java/org/eclipse/ice/dev/annotations/processors/VelocitySourceWriter.java

```diff
@@ -28,23 +31,42 @@
 
 	protected String template;
 	protected Map<String, Object> context;
+	protected JavaFileObject generatedFile;
 	
 	public VelocitySourceWriter() {
 		this.context = new HashMap<>();
 	}
-
+	
+	/**
+	 * Opens a writer and then proceed to write a source file
+	 * @param generatedFile
+	 * @throws IOException
+	 */
+	public void write(JavaFileObject generatedFile) throws IOException {
+		try (Writer writer = generatedFile.openWriter()) {
+			write(writer);
+		}
+	}
+	
```

Comment:
I think we have a problem here with the `SourceWriter`s knowing the implementation detail of its `Writer` being opened from `JavaFileObject`. This is something we missed when we talked about this; we said we could just instantiate the `SourceWriter` with a `Writer` but when we actually want to write the file, the `Writer` would be closed, as we found out with that elusive bug lol... I don't have a solution but wanted to call this out as a potential point of discussion

Reply:
I think it would be better to have write() check to see if the writer is open, open it if not, and then write. 

If that won't work, then it is probably fine to document the requirement that the writer be open. Note that your current documentation says that it will open the writer, but clearly it doesn't.

Reply:
We could perhaps wrap it with a lambda if you dont like sourcewriter haveing access to JavaFileObjectss.
e.g.
Supplier<Writer> writerSupplier = () -> generatedFile.openWriter();

So that inside the SourceWriter it will only have access to a lambda that produces a writer.

Reply:
From the [standard `Writer` interface](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/io/Writer.html), it looks like reopening or even checking if a writer is open isn't an option, unfortunately

Reply:
@jayjaybillings This version does in fact open a writer.  Dan was referring to a bug I had faced previously where I was opening and storing writers, but they would close before the write file portion of the code.  My remedy was to wait till we needed to do the writing, and then open/create the writer with the JavaFileObject.

Reply:
> We could perhaps wrap it with a lambda if you dont like sourcewriter haveing access to JavaFileObjectss.
> e.g.
> Supplier writerSupplier = () -> generatedFile.openWriter();
> 
> So that inside the SourceWriter it will only have access to a lambda that produces a writer.

So something like:
```java
protected Supplier<Writer> writerSupplier;
/**
 * Opens a writer and then proceed to write a source file
 * @param generatedFile
 * @throws IOException
 */
public void write() throws IOException {
    try (Writer writer = writerSupplier.get()) {
            write(writer);
    }
}
```

That seems reasonable to me :+1:

Reply:
@walshmm Right - I see that now, thanks. 

@dbluhm What implementation detail is the problem here?

Reply:
@jayjaybillings `JavaFileObject` being known to the `VelocitySourceWriter`. I think it breaks down the separation of concerns and makes it less reusable for it to open writers from `JavaFileObject`s directly

Reply:
Yeah, but it's below the code smell limit for me. I think it's fine for now.